File: codility/src/main/java/codility/com/lessons/l09maxsliceproblem/MaxProfit.java
package codility.com.lessons.l09maxsliceproblem;

/**
 * 100%
 * https://app.codility.com/programmers/lessons/9-maximum_slice_problem/max_profit/
 */
public class MaxProfit {
	public int solution(int[] A) {
		if (A.length < 2) {
			return 0;
		}

		int max_ending = 0;
		int max_slice = 0;
		int min = A[0];
		for (int i = 1; i < A.length; i++) {
			max_ending = Math.max(0, A[i] - min);
			min = Math.min(min, A[i]);
			max_slice = Math.max(max_slice, max_ending);
		}
		return max_slice;
	}
}

File: codility_new/max_profit.py
# https://app.codility.com/programmers/lessons/9-maximum_slice_problem/max_profit/

def solution(A):
    if len(A) == 0:
        return 0
    min_so_far = [A[0]] * (len(A) + 1)
    # max_so_far = [A[0]] * (len(A) + 1)

    for i in range(len(A)):
        min_so_far[i + 1] = min(min_so_far[i], A[i])
        # max_so_far[i + 1] = max(max_so_far[i], A[i])
    
    max_profit = 0
    for i in range(len(A)):
        max_profit = max(max_profit, A[i] - min_so_far[i + 1])

    return max_profit
